ktor-client/ktor-client-core/common/src/io/ktor/client/features/NeedRetry.kt

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2014-2019 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.client.features
+
+import io.ktor.client.HttpClient
+import io.ktor.client.HttpClientConfig
+import io.ktor.client.request.HttpRequestBuilder
+import io.ktor.client.request.takeFrom
+import io.ktor.client.response.HttpReceivePipeline
+import io.ktor.client.response.HttpResponse
+import io.ktor.util.AttributeKey
+
+/**
+ * Need retry condition checker
+ *
+ * You could return true to make [HttpClient] retry the request
+ */
+typealias RetryCondition = suspend (requestBuilder: HttpRequestBuilder, response: HttpResponse) -> Boolean
+
+/**
+ * Need retry feature for [HttpClient]
+ *
+ * @property retryHandlers: list of retry conditions to check before deciding to retry a request
+ */
+class NeedRetry(
+    private val retryHandlers: List<RetryCondition>
+) {
+
+    /**
+     * [NeedRetry] configuration.
+     */
+    class Config {
+        internal val retryHandlers: MutableList<RetryCondition> = mutableListOf()
+
+        /**
+         * Add [RetryCondition].
+         * Last added handler executes first.
+         */
+        fun retryCondition(block: RetryCondition) {
+            retryHandlers += block
+        }
+    }
+
+    companion object : HttpClientFeature<Config, NeedRetry> {
+        override val key: AttributeKey<NeedRetry> = AttributeKey("NeedRetry")
+
+        override fun prepare(block: Config.() -> Unit): NeedRetry {
+            val config = Config().apply(block)
+            return NeedRetry(config.retryHandlers)
```

Comment:
```suggestion
            return NeedRetry(config.retryHandlers.reversed())
```
Should reverse the list because line 39 mentions "Last added handler executes first.".